tests/storage/test_input.py

```diff
@@ -92,6 +92,9 @@ def _from_hdf(self, hdf, version=None):
 class CaffeinatedTrait(TraitType):
     """
     We can even make our own trait types.
+
+    In this case, we're making a counter-example -- custom traits with mutable defaults are dangerous! If more than one
+    object uses this trait, their default is the *same instance* of the mutable default.
```

Comment:
This is true for _all_ class or function definitions with mutable defaults! That's one reason for the importance of 
```
def some(arg0=None):
   arg0 = arg0 or {'default': True}
```
pattern.

Reply:
Yeah, this is a great point. I think in my head I was so impressed with traitlets that I figured they were doing some sort of magic in `__new__` with `inspect` or somesuch to get the keyword defaults to re-instantiate as fresh objects? But you're right that this is totally expected; I'll re-phrase the documentation to emphasize how the `@default` decorator is the way to get around this expected behaviour.

Reply:
I'm glad you pushed me on this, because it turns out that you totally can define a custom `TraitType` with a mutable default! It's not covered in their [readthedocs](https://traitlets.readthedocs.io/en/stable/defining_traits.html), but poking around the source I discovered that all you need to do is define a `make_dynamic_default(self)` method.